pkg/monitor/shutdown/consumer.go

```diff
@@ -0,0 +1,160 @@
+package shutdown
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/openshift/origin/pkg/monitor/monitorapi"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	Reason                       = "GracefulShutdownWindow"
+	KubeAPIServerShutdownLocator = "shutdown/apiserver"
+)
+
+var (
+	namespaceToServer = map[string]string{
+		"openshift-kube-apiserver": "kube-apiserver",
+		"openshift-apiserver":      "openshift-apiserver",
+		// TODO: oauth apiserver does not seem to emit any shutdown events
+		"openshift-oauth-apiserver": "oauth-apiserver",
+	}
+)
+
+// Monitor abstracts the monitor API
+type Monitor interface {
+	StartInterval(t time.Time, condition monitorapi.Condition) int
+	EndInterval(startedInterval int, t time.Time)
+}
+
+func newConsumer(monitor Monitor) *consumer {
+	return &consumer{
+		monitor:   monitor,
+		byHost:    make(map[string]*corev1.Event),
+		processed: make(map[types.UID]struct{}),
+	}
+}
+
+type consumer struct {
+	monitor   Monitor
+	lock      sync.Mutex
+	processed map[types.UID]struct{}
+	byHost    map[string]*corev1.Event
+}
+
+func (c *consumer) Consume(event *corev1.Event) {
+	if event == nil || !interesting(event) {
+		return
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	var ok bool
+	if _, ok := c.processed[event.UID]; ok {
+		return
+	}
+
+	key := key(event)
+	var start, end *corev1.Event
+	start, ok = c.byHost[key]
+	if !ok {
+		c.byHost[key] = event
+		return
+	}
+	end = event
+
+	delete(c.byHost, key)
+	c.processed[start.UID] = struct{}{}
+	c.processed[end.UID] = struct{}{}
+
+	condition := complete(start, end)
+	// these logs can be useful to debug if the windows are displayed correctly
+	framework.Logf("GracefulShutdownEvent - Start: %+v", start)
```

Comment:
Should this be logged when start is collected to maintain correct timestamp?

Reply:
I am logging the pair next to each other so we can see how the window was constructed, and the interval timestamp should still be correct since we take the time from the  'CreationTimestamp' field of the event.